### ISSUE TITLE
insiders > 4.7.0 fails with plugin mkdocs-exclude-search

### ISSUE THREAD

**Issue by lampensau** [squidfunk/mkdocs-material#3543]
### Contribution guidelines

- [X] I've read the [contribution guidelines](https://github.com/squidfunk/mkdocs-material/blob/master/CONTRIBUTING.md) and wholeheartedly agree

### I've found a bug and checked that ...

- [X] ... the problem doesn't occur with the `mkdocs` or `readthedocs` themes
- [X] ... the problem persists when all overrides are removed, i.e. `custom_dir`, `extra_javascript` and `extra_css`
- [X] ... the documentation does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem

### Description

With offline search introduced in insiders 4.7.0 I'm no longer able to build our documentation because I'm using the plugin [mkdocs-exclude-search](https://github.com/chrieke/mkdocs-exclude-search).

This problem is also present when not explicitly having mkdocs-material defined in `mkdocs.yml`. It is enough to have the (insiders) theme installed.

### Expected behaviour

The build should complete successfully as with versions < 4.7.0.

### Actual behaviour

A build fails with the following error:

```
INFO     -  Cleaning site directory
INFO     -  Building documentation to directory: /home/user/my-docs/testing/site
Traceback (most recent call last):
  File "/home/user/my-docs/testing/venv/bin/mkdocs", line 8, in <module>
    sys.exit(cli())
  File "/home/user/my-docs/testing/venv/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/user/my-docs/testing/venv/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/user/my-docs/testing/venv/lib/python3.9/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/user/my-docs/testing/venv/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/user/my-docs/testing/venv/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/user/my-docs/testing/venv/lib/python3.9/site-packages/mkdocs/__main__.py", line 187, in build_command
    build.build(config.load_config(**kwargs), dirty=not clean)
  File "/home/user/my-docs/testing/venv/lib/python3.9/site-packages/mkdocs/commands/build.py", line 317, in build
    config['plugins'].run_event('post_build', config=config)
  File "/home/user/my-docs/testing/venv/lib/python3.9/site-packages/mkdocs/plugins.py", line 104, in run_event
    result = method(**kwargs)
  File "/home/user/my-docs/testing/venv/lib/python3.9/site-packages/mkdocs_exclude_search/plugin.py", line 266, in on_post_build
    with open(search_index_fp, "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: '/home/user/my-docs/testing/site/search/search_index.json'
```

### Steps to reproduce

1. Install mkdocs, mkdocs-exclude-search, and mkdocs-material-insiders in a new python venv
2. Create a new project `mkdocs new testing`
3. Modify `mkdocs.yml` to include mkdocs-exclude-search with `exclude:` property.
4. Build or serve documentation

### Package versions

- Python: `Python 3.9.2`
- MkDocs: `1.2.3`
- Material: `8.1.10+insiders.4.8.0`

```
$ pip list | grep mkdocs
mkdocs                     1.2.3
mkdocs-exclude-search      0.6.4
mkdocs-material            8.1.10+insiders.4.8.0
mkdocs-material-extensions 1.0.3
```

### Configuration

```yaml
plugins:
  - search:
      lang: en
  - exclude-search:
      exclude:
        - index.md
```

### System information

- Operating system: Debian 11 (Bullseye)

**Comment by squidfunk** [comment]
Confirmed. That's very much not intended. I'll take a look and see how we can keep the offline capability while allowing for third-party plugins to alter the search index.

**Comment by squidfunk** [comment]
Okay, I think I have a working version in [`98201fd`](https://github.com/squidfunk/mkdocs-material-insiders/commit/98201fd4a3e401258096969da0cb3434f6577280). I've factored the offline support out into a separate plugin which can be added at the end of your configuration. If this plugin is enabled, it will make the necessary transformations for offline support, meaning that the search index might be transformed by other plugins _before_ it is picked up by this plugin:

``` yaml
plugins:
  - search
  - exclude-search
  - offline
```

It will also automatically set `use_directory_urls: false`, as this is a mandatory setting for offline documentation. I've yet to do some cleanup, but the plugin is already in the `master` of Insiders and can be used as described above. You can now build documentation for both environments by enabling the plugin, but this will also mean that your online build will have `index.html` endings in URLs. Thus, you can disable the plugin for online builds, and enable it for offline builds. I'll add a simple configuration setting when I do the cleanup, so it can be set via environment variable, like so:

``` yaml
plugins:
  - offline:
      enable: !ENV OFFLINE
```

Same for documentation. Just wanted to let you know that there's a fix in-place. Together with the new privacy plugin, which pulls down all external assets it's a powerful combination to build completely self-contained offline documentation.

**Comment by squidfunk** [comment]
Okay, I think I have a working version in [`98201fd`](https://github.com/squidfunk/mkdocs-material-insiders/commit/98201fd4a3e401258096969da0cb3434f6577280). I've factored the offline support out into a separate plugin which can be added at the end of your configuration. If this plugin is enabled, it will make the necessary transformations for offline support, meaning that the search index might be transformed by other plugins _before_ it is picked up by this plugin:

``` yaml
plugins:
  - search
  - exclude-search
  - offline
```

It will also automatically set `use_directory_urls: false`, as this is a mandatory setting for offline documentation. I've yet to do some cleanup, but the plugin is already in the `master` of Insiders and can be used as described above. You can now build documentation for both environments by enabling the plugin, but this will also mean that your online build will have `index.html` endings in URLs. Thus, you can disable the plugin for online builds, and enable it for offline builds. I'll add a simple configuration setting when I do the cleanup, so it can be set via environment variable, like so:

``` yaml
plugins:
  - offline:
      enable: !ENV OFFLINE
```

Same for documentation. Just wanted to let you know that there's a fix in-place. Together with the new privacy plugin, which pulls down all external assets it's a powerful combination to build completely self-contained offline documentation.

**Comment by squidfunk** [comment]
The new [built-in offline plugin](https://squidfunk.github.io/mkdocs-material/setup/building-for-offline-usage/) was just released as part of [insiders-4.10.0](https://squidfunk.github.io/mkdocs-material/insiders/changelog/#4.10.0) 😊 As mentioned in [my last comment](https://github.com/squidfunk/mkdocs-material/issues/3543#issuecomment-1052177462), this fixes the problem with third-party plugins altering the search index, which could not be used since insiders 4.8.0 – a design flaw which was an oversight on my side. I apologize for the inconvenience this was causing you (and possibly others).
